lib/mongo/server/connection.rb

```diff
@@ -190,9 +190,9 @@ def handshake!
 
       def authenticate!
         if options[:user] || options[:auth_mech]
-          user = Auth::User.new(Options::Redacted.new(:auth_mech => default_mechanism).merge(options))
+          @user ||= Auth::User.new(Options::Redacted.new(:auth_mech => default_mechanism).merge(options))
```

Comment:
From looking at the other authentication mechanisms, none of them seem to cache any state from the authentication conversation, so I think it should be fine to cache the user without checking that the mechanism is SCRAM.